src/bloqade/pyqrack/qasm2/core.py

```diff
@@ -77,3 +80,18 @@ def creg_eq(
             return (False,)
 
         return (all(left is right for left, right in zip(lhs, rhs)),)
+
+    @interp.impl(glob.UGate)
+    def ugate(self, interp: PyQrackInterpreter, frame: interp.Frame, stmt: glob.UGate):
+        registers: ilist.IList[PyQrackReg, Any] = frame.get(stmt.registers)
+        theta, phi, lam = (
+            frame.get(stmt.theta),
+            frame.get(stmt.phi),
+            frame.get(stmt.lam),
+        )
+
+        for qreg in registers:
+            for qarg in qreg:
```

Comment:
You can also implement the `__iter__` magic method for `PyQrackReg` to get the correct behavior here.